NXP: Add support for Kinetis KE1xF SoC series and TWR-KE18F development board

henrikbrixandersen opened the pull request:
This patch series adds support for the NXP Kinetis KE1xF SoC series and the TWR-KE18F development board through the following commits:

* boards: twr-ke18f: add support for the NXP TWR-KE18F board
* soc: arm: ke1xf: add NXP Kinetis KE1xF SoC series support
* drivers: clock_control_mcux_sim: add support for clkout
* drivers: serial: uart_mcux_lpuart: add support for LPUARTs with two IRQs
* clock_control: introduce mcux pcc driver
* ext: hal: nxp: mcux: add device files for KE1xF series SoCs

henrikbrixandersen committed:
ext: hal: nxp: mcux: add device and driver files for KE1xF series SoCs

Add mcux 2.5.0 device and driver files for NXP Kinetis KE1xF series
SoCs (MKE14F16, MKE16F16, and MKE18F16).

Origins: NXP MCUxpresso SDK 2.5.0
URL: mcuxpresso.nxp.com
Maintained-by: External

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
clock_control: introduce mcux pcc driver

Add a new clock control driver for NXP Kinetis SoCs that have the
Peripheral Clock Controller module (PCC).

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
soc: arm: ke1xf: add NXP Kinetis KE1xF SoC series support

Add initial support for the NXP Kinetis KE1xF SoC series (MKE14F16,
MKE16F16, and MKE18F16).

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: serial: uart_mcux_lpuart: add support for LPUARTs with two IRQs

Add support for NXP MCUX LPUART devices with separate IRQ lines for
transmit and receive status interrupts (e.g. the Kinetis KE1xF SoC
series).

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
boards: twr-ke18f: add support for the NXP TWR-KE18F board

Add support for the NXP TWR-KE18F development board. This board
feautures an NXP MKE18F16 MCU, a selection of user LEDs and
push-buttons, potentiometer, thermistor, CAN interface, and FlexIO
header.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
PR adding support for further peripherals (SPI, I2C, WDOG, CAN, ADC12, ...) will follow once ready.

dbkinder commented on boards/arm/twr_ke18f/doc/index.rst:
```diff
@@ -0,0 +1,211 @@
+Overview
+********
+
+The TWR-KE18F is a development board for platform for the NXP Kinetis
```
How about:
```
The TWR-KE18F is a development board for NXP Kinetis
KE1xF 32-bit MCU-based platforms.  The onboard OpenSDAv2 serial and
debug adapter, running an open source bootloader, offers options for
serial communication, flash programming, and run-control debugging.
```

henrikbrixandersen replied:
@dbkinder I appreciate your review. I have made the suggested changes.

dbkinder commented on boards/arm/twr_ke18f/doc/index.rst:
```diff
@@ -0,0 +1,211 @@
+- CAN pin header
+- Flex I/O pin header
+
+For more information about the KE1xF SoC and the TWR-KE18F board:
```
```suggestion
For more information about the KE1xF SoC and the TWR-KE18F board,
see these NXP reference documents:
```

dbkinder commented on boards/arm/twr_ke18f/doc/index.rst:
```diff
@@ -0,0 +1,211 @@
+| FLASH     | on-chip    | soc flash                           |
++-----------+------------+-------------------------------------+
+
+The default configuration can be found in the defconfig file:
```
This renders better as:
```
The default configuration can be found in the defconfig file:
``boards/arm/twr_ke18f/twr_ke18f_defconfig``.
```

dbkinder reviewed (changes requested):
Doc changes look good, with a few suggested edits

henrikbrixandersen commented on ext/hal/nxp/mcux/README:
```diff
@@ -24,6 +24,9 @@ Status:
    MK64F12	SDK 2.5.0 (2018-12-17)		REL_2.5.0_REL9_RFP_RC3_7_1
+   MKE14F16	SDK 2.5.0 (2018-12-17)		REL_2.5.0_REL9_RFP_RC3_7_1
+   MKE16F16	SDK 2.5.0 (2018-12-17)		REL_2.5.0_REL9_RFP_RC3_7_1
+   MKE18F16	SDK 2.5.0 (2018-12-17)		REL_2.5.0_REL9_RFP_RC3_7_1
```
@MaureenHelm I am not sure if the tag is correct for these imports. Can I see the tag from the generated MCUX SDK archive somehow?

henrikbrixandersen replied:
<img width="861" alt="Screenshot 2019-05-10 at 19 37 26" src="https://user-images.githubusercontent.com/1076226/57545815-bc4ea680-735b-11e9-81bd-15f7809db89d.png">

@MaureenHelm My list of recent SDKs does not mention the tag. Perhaps I should leave them blank (or `-` in the `README` file?

MaureenHelm replied:
Let me check with the SDK team to find out why the tag isn't there

MaureenHelm replied:
It turns out that only employee accounts can see the tag. Instead, please use the "revision" attribute from `TWR-KE18F_manifest_v3_4.xml`:
```xml
<ksdk id="MCUXpresso250" name="MCUXpresso250" version="2.5.0" revision="150 2018-12-17"/>
```

henrikbrixandersen replied:
@MaureenHelm Thank you for following up on this. I have changed the `README` file to list the `revision` attribute values.

henrikbrixandersen committed:
drivers: clock_control_mcux_sim: add support for clkout

Add support for CLKOUT source selection and divider as found on the
NXP Kinetis KE1xF SoC series.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

MaureenHelm reviewed (changes requested):
Wow, great work! Just a couple of nits, and please confirm the SDK version.

henrikbrixandersen commented on drivers/clock_control/clock_control_mcux_pcc.c:
```diff
@@ -0,0 +1,78 @@
+};
+
+#if defined(DT_MCUX_PCC_0_NAME)
+static struct mcux_pcc_config mcux_pcc0_config = {
```
Thank you for the swift review. I have made this const, good catch.

henrikbrixandersen commented on dts/arm/nxp/nxp_ke1xf.dtsi:
```diff
@@ -0,0 +1,186 @@
+		};
+
+		uart2: uart@4006c000 {
+			compatible = "nxp,kinetis-uart";
```
Fixed, thanks.

dbkinder reviewed (approved):
+1 for docs, thanks!
